### PR TITLE
wtf8, char: Replace uses of `mem::transmute` with more specific functions

### PR DESCRIPTION
**Pull request by tbu-**

### CODE REVIEW

File: src/libcore/char.rs [hunk 1]
```diff
@@ -84,10 +84,17 @@ pub fn from_u32(i: u32) -> Option<char> {
...
     } else {
-        Some(unsafe { transmute(i) })
+        Some(unsafe { from_u32_unchecked(i) })
     }
 }
 
+/// Converts a `u32` to an `char`, not checking whether it is a valid unicode
+/// codepoint.
+#[inline]
+pub unsafe fn from_u32_unchecked(i: u32) -> char {
```
Comment (alexcrichton) [inline src/libcore/char.rs]: Can you add an explicit `#[unstable]` annotation to this function? It should have its own feature name as well.

**Comment by alexcrichton** [comment]
@bors: r-

Just wanna make sure that new public API surface area is documented with `#[unsafe]`

**Comment by tamird** [comment]
s/unsafe/unstable/

**Comment by tbu-** [comment]
@alexcrichton 
Done.

File: src/libcore/char.rs [hunk 2]
```diff
@@ -84,10 +84,18 @@ pub fn from_u32(i: u32) -> Option<char> {
...
     } else {
-        Some(unsafe { transmute(i) })
+        Some(unsafe { from_u32_unchecked(i) })
     }
 }
 
+/// Converts a `u32` to an `char`, not checking whether it is a valid unicode
+/// codepoint.
+#[inline]
+#[unstable(feature = "char_from_unchecked")]
```
Comment (alexcrichton) [inline src/libcore/char.rs]: Could you also add a `reason` here along the lines of `reason = "recently added API"` ? Helps us keep track as to *why* it's unstable.

**Comment by tbu-** [comment]
@alexcrichton 
Added a `reason`.

**Pull request merged**